zookeeper-api/src/main/java/org/apache/helix/zookeeper/datamodel/serializer/ZNRecordSerializer.java

```diff
@@ -81,25 +82,48 @@ private static int getListFieldBound(ZNRecord record) {
     serializationConfig.set(SerializationConfig.Feature.AUTO_DETECT_FIELDS, true);
     serializationConfig.set(SerializationConfig.Feature.CAN_OVERRIDE_ACCESS_MODIFIERS, true);
     ByteArrayOutputStream baos = new ByteArrayOutputStream();
-    byte[] serializedBytes;
+    byte[] serializedBytes = new byte[0];
+    boolean isCompressed = false;
+
     try {
       mapper.writeValue(baos, data);
       serializedBytes = baos.toByteArray();
       // apply compression if needed
-      if (record.getBooleanField("enableCompression", false) || serializedBytes.length > ZNRecord.SIZE_LIMIT) {
+      if (ZNRecordUtil.shouldCompress(record, serializedBytes.length)) {
         serializedBytes = GZipCompressionUtil.compress(serializedBytes);
+        isCompressed = true;
       }
     } catch (Exception e) {
-      logger.error("Exception during data serialization. Will not write to zk. Data (first 1k): "
-          + new String(baos.toByteArray()).substring(0, 1024), e);
+      if (serializedBytes.length == 0 || GZipCompressionUtil.isCompressed(serializedBytes)) {
+        serializedBytes = baos.toByteArray();
```

Comment:
This is a potential bug, "serializedBytes = baos.toByteArray();" is called in the try block. As we discussed, either split the try block, or just don't try to read the content in catch block.

Reply:
Removed the logs.